### DISCUSSION TITLE
Where in the codebase are the outputs of TctColBertDocumentEncoder/TctColBertQueryEncoder combined to create the MaxSim score?

### DISCUSSION THREAD

**Discussion by plamb-viso** [castorini/pyserini#1646] [Q&A] [answered]
`TctColBertDocumentEncoder/TctColBertQueryEncoder` make it super easy to setup document/query encoding with `castorini/tct_colbert-v2-hnp-msmarco` but I cannot find where the outputs from the `encode` functions in these classes are combined to create a MaxSim score per document.

**Comment by jacklin64** [comment, marked as answer]
Hi @plamb-viso, thanks for the question. Here we'd like to clarify that TCTColBERT uses dot product operation to compute scores like standard single-vector DPR  instead of ColBERT using MaxSim. The main difference between TCTColBERT and DPR is training, where TCTColBERT learns knowledge from ColBERT maxim scores through all the in-batch samples but after training. See more detail in our paper https://aclanthology.org/2021.repl4nlp-1.17/

**Reply by plamb-viso** [reply]
Ah so the final output between `TctColBertDocumentEncoder/TctColBertQueryEncoder` would be a dot product? Curious if there is an example, say, in a test case, where this is computed for the outputs of these two classes `encode` method in the codebase. Basically what I'm interested in is an example of encoding a couple documents, encoding a query and computing the final score per each document.

**Reply by jacklin64** [reply]
I don't remember we have script to directly compute inner product between TCT-ColBERT encoded query and documents. But you can use the script here to compute the embedding for documents https://github.com/castorini/pyserini/blob/db1347c40ab1840f1e0851312ae3999d76f98228/scripts/tct_colbert/encode_corpus_msmarco_doc.py#L107. And also creating a `TctColBertQueryEncoder` class and encode query embedding.

**Reply by plamb-viso** [reply]
Ah I should've stated what i was trying to do -- I'm using ColBERT for reranking, so it gets a set of documents and a query on the fly; those documents are dynamic so you cant index/precompute them beforehand. That's why I'm interested in an example of computing the inner product (it looks like perhaps here FAISS will be computing it over previously indexed documents).

**Reply by plamb-viso** [reply]
Would you suggest just using a [numpy.inner()](https://numpy.org/doc/stable/reference/generated/numpy.inner.html) or [numpy.dot()](https://numpy.org/doc/stable/reference/generated/numpy.dot.html) ?

**Reply by plamb-viso** [reply]
hi @jacklin64 just wanted to try pinging you one more time on the above
